File: golang/_0015_3Sum/main.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	res, n := make([][]int, 0), len(nums)
	if n > 2 {
		sort.Ints(nums)
		for i := 0; i < n-2; i++ {
			if i > 0 && nums[i] == nums[i-1] {
				continue
			}

			for j, k := i+1, n-1; j < k; {
				sum := nums[j] + nums[k] + nums[i]
				if sum == 0 {
					res = append(res, []int{nums[i], nums[j], nums[k]})
					j++
					for j < k && nums[j] == nums[j-1] {
						j++
					}
				} else if sum > 0 {
					k--
				} else {
					j++
				}
			}
		}
	}

	return res
}

File: python3/_0015_3Sum/main.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = list()

        n = len(nums)
        if n >= 3:
            nums.sort()

            for i in range(n):
                if i > 0 and nums[i] == nums[i - 1]:
                    continue

                j, k = i + 1, n - 1
                while j < k:
                    if nums[i] + nums[j] + nums[k] == 0:
                        res.append([nums[i], nums[j], nums[k]])
                        j += 1
                        while j < k and nums[j] == nums[j - 1]:
                            j += 1
                    elif nums[i] + nums[j] + nums[k] > 0:
                        k -= 1
                    else:
                        j += 1

        return res
